js/client-mainmenu.js

```diff
@@ -114,7 +131,118 @@
 				);
 			}
 		},
+		createLoginFormObserver:function(self){
+			console.log(window.Storage.prefs("page"))
+			const config = { attributes: true, childList: true, subtree: true };
+			const observer = new MutationObserver(function(){
+			if(window.app.user.get("named") && window.Storage.prefs("page")=="challenge"){
+				window.Storage.prefs("page","testclient");
+				if($('.challenge-container').length==0){
+					//user log in event trigerd when user log in and make user pick a team
+					$('body').on("login_event",function(){
+					self.onUserLogIn(self);
+				})
+				$('body').trigger("login_event");
+				}
+			}
+			});
+			observer.observe(document.activeElement, config);
+			},
+		onUserLogIn:function(self){
+			var buf = '<div class="challenge"><form class="battleform"><p>Challenge from ' + BattleLog.escapeHTML(window.Storage.prefs("challenger")) + '?</p>';
+			buf += '<p><label class="label">Format:</label>' + self.renderFormats(window.Storage.prefs("format")) + '</p>';
+			buf += '<p id="select-team"><label class="label">Team:</label>' + self.renderTeams(window.Storage.prefs("format")) + '</p>';
+			buf +='<button type="button" id=accept-challenge>accept</button>';
+			buf +='<button type="button" id=close-challenge>close</button></form></div>'
+			$('.pmbox').append(`<div class=challenge-container>${buf}</div>`);
+			$('.select.formatselect').prop("disabled",true);
+			$('.select.teamselect').prop("disabled",false);
+			$('#accept-challenge').click(function(){
+				self.onTeamSelect(window.Storage.prefs("challenger")).then(function(data){
+					 $('#accept-challenge').html("accept");
+					if(data.massage.includes("Your team is valid") && data.userId!=window.app.user.get("name") && data.userId){
+						self.sendChallenge({userId:data.userId,format:$('.select.formatselect').val()});
+						$('.challenge-container').empty();
+					 }
+					 else if(data.userId==app.user.get("name")){
+						window.app.addPopupMessage("opponent name and challeger name  are equall you cant challenge your self");
+					 }
+					else{
+						window.app.addPopupMessage(data.massage);
+					}
+				})
+			})
+			$('#close-challenge').click(function(){
+				$('.challenge-container').empty();
+			})
+		},
+		onTeamSelect:function(challengerName){
+			//used promise to wait for reponde for team pick
+			return new Promise(function(resolve){
```

Comment:
A callback is probably preferred over a Promise. I'm not sure we polyfill that here, and callback is the general style in the rest of the code.

Reply:
do I change promise to callback function or leave it as it is

Reply:
i think promise is better that way i can assure the client will get the correct message

Reply:
A callback function is preferable (no compatibility issues, same style as the rest of the code.)